src/embedded/device_info.cpp

```diff
@@ -0,0 +1,137 @@
+#include "device_info.h"
+#include <QFile>
+#include <QTextStream>
+#include <QJsonObject>
+#include <QDebug>
+
+class DeviceInfo::impl{
+    bool _is_raspberry_pi = false;
+    QString _revision = "";
+    QString _device_code = "";
+    QJsonArray _tags = {};
+    bool _tags_set = false;
+    QString _hardware_name = "";
+   
+public:
+    void determineHardware()
+    {
+        // Attempt to open /proc/cpuinfo
+        QFile f("/proc/cpuinfo");
+        if (f.open(QFile::ReadOnly | QFile::Text)) {
+            QTextStream s(&f);
+            while (!s.atEnd()) {
+                QString line = s.readLine();
+                if (line.startsWith("Revision")) {
+                    _revision = line.section(':', 1).trimmed();
+                    bool ok;
+                    uint32_t value_int = _revision.toUInt(&ok, 16);
+                    if (!ok) { qDebug() << "Revision Code Not OK!"; }
+
+                    // New Style Revision Code
+                    // Device type is 8 bits, stored in bits 4-11
+                    const uint32_t MASK = 0xFF << 4;
+                    uint32_t isolated_bits = value_int & MASK;
+                    uint32_t extracted_value = isolated_bits >> 4;
+
+                    // convert final value back to HEX
+                    std::stringstream ss;
+                    ss << std::hex << extracted_value;
+                    QString extracted_hex_string = QString::fromStdString(ss.str());
+                    _device_code = extracted_hex_string;

```

Comment:
I think I'd be tempted to store the `_device_code` as the actual integer, and only convert to hex-string when needed, rather than storing `_device_code` as a `QString`? But of course that's entirely up to you.